### ISSUE TITLE
`groups` parameter for `nn.Linear`

### ISSUE THREAD

**Issue by moi90** [pytorch/pytorch#115852]
### 🚀 The feature, motivation and pitch

[nn.Conv2d](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html#conv2d) has a parameter groups:

    groups controls the connections between inputs and outputs. in_channels and out_channels must both be divisible by groups. For example,
    - At groups=1, all inputs are convolved to all outputs.
    - At groups=2, the operation becomes equivalent to having two conv layers side by side

I think, something similar makes sense for [nn.Linear](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html#torch.nn.Linear). For example, it would have the advantage of reducing the number of parameters.

### Alternatives

This is what I came up with:

```python
class GroupedLinear(nn.Module):
    def __init__(
        self,
        in_features: int,
        out_features: int,
        bias: bool = True,
        groups=1,
        device=None,
        dtype=None,
    ) -> None:
        super().__init__()

        assert in_features % groups == 0
        assert out_features % groups == 0

        self.in_features = in_features
        self.out_features = out_features
        self.groups = groups

        self._linear_layers = nn.ModuleList(
            [
                nn.Linear(
                    in_features // groups,
                    out_features // groups,
                    bias=bias,
                    device=device,
                    dtype=dtype,
                )
                for _ in range(groups)
            ]
        )

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = x.view((x.shape[0], self.groups, -1))
        result = [
            l(x[:, i])
            for i, l in enumerate(self._linear_layers)
        ]
        return torch.cat(result, dim=1)
```

(But I don't know if it is correct and performant.)

### Additional context

I asked this already on [discuss.pytorch.org](https://discuss.pytorch.org/t/groups-parameter-for-nn-linear/182036/1) but received no reply.

**Comment by andreasfloros** [comment]
A linear layer is equivalent to convolution with `kernel_size=1`. Since convolutions already support groups, you can rewrite this as a single `Conv1d` layer. For example, the following is equivalent to the implementation you gave:

```python

import torch
import time


class GroupedLinear(torch.nn.Module):
    def __init__(
        self,
        in_features: int,
        out_features: int,
        bias: bool = True,
        groups=1,
        device=None,
        dtype=None,
    ) -> None:
        super().__init__()

        assert in_features % groups == 0
        assert out_features % groups == 0

        self.in_features = in_features
        self.out_features = out_features
        self.groups = groups

        self._linear_layers = torch.nn.ModuleList(
            [
                torch.nn.Linear(
                    in_features // groups,
                    out_features // groups,
                    bias=bias,
                    device=device,
                    dtype=dtype,
                )
                for _ in range(groups)
            ]
        )

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = x.view((x.shape[0], self.groups, -1))
        result = [
            l(x[:, i])
            for i, l in enumerate(self._linear_layers)
        ]
        return torch.cat(result, dim=1)


class GroupedConv1d(torch.nn.Module):
    def __init__(
        self,
        in_features: int,
        out_features: int,
        bias: bool = True,
        groups=1,
        device=None,
        dtype=None,
    ) -> None:
        super().__init__()
        self._conv = torch.nn.Conv1d(in_channels=in_features,
                                     out_channels=out_features,
                                     kernel_size=1,
                                     groups=groups,
                                     bias=bias,
                                     device=device,
                                     dtype=dtype)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = x.unsqueeze(2)
        return self._conv(x).squeeze(2)


in_features = 20000
out_features = 40000
groups = 500
batch_size = 8
x = torch.randn(batch_size, in_features)
linear = GroupedLinear(in_features, out_features, groups=groups)
start = time.perf_counter()
y_linear = linear(x)
print("GroupedLinear time:", time.perf_counter() - start)

conv = GroupedConv1d(in_features, out_features, groups=groups)

weights = torch.cat([layer.weight
                     for layer in linear._linear_layers], dim=0).unsqueeze(2)
biases = torch.cat([layer.bias for layer in linear._linear_layers], dim=0)
with torch.no_grad():
    conv._conv.weight.copy_(weights)
    conv._conv.bias.copy_(biases)

start = time.perf_counter()
y_conv = conv(x)
print("GroupedConv time:", time.perf_counter() - start)

print("Error:", torch.norm(y_linear - y_conv))

```

This is probably a bit more elegant and avoids looping in the forward.

**Comment by moi90** [comment]
I resolved it like this:

```python
def grouped_linear(input, weight, bias=None, groups=1):
    N = input.size(0)
    # Divide channels into groups and swap groups with batch dimension
    input = input.view(N, groups, -1).transpose(1, 0)

    # Swap groups axis with batch dimension and remove
    output = torch.matmul(input, weight)

    output = output.transpose(1, 0).reshape(N, -1)

    if bias is not None:
        output += bias

    return output
```

However, I did not do a performance comparison. (The `.reshape(N, -1)` in the end probably reduces the performance.)
